### ISSUE TITLE
Problems with I/O of Arrays with replicated dimensions

### ISSUE THREAD

**Issue by billsacks** [esmf-org/esmf#184]
In doing some testing of #145 - specifically, testing the case that combines (a) multi-tiles, (b) more than 1 DE per PET, and (c) replicated dimensions - I noticed some oddities with I/O of Arrays with replicated dimensions. I have reproduced these oddities in a test case with single-tile Arrays: see https://github.com/billsacks/esmfprojects-replicated_dims_io (testing was done with billsacks/esmfprojects-replicated_dims_io@b96f2527520d43da452502b41248488d76a59713 and ESMF at the head of develop, 45807aad7a).

In a simple test case with 1 DE per PET, I get an output file that contains the replicated dimensions, but all slices of the replicated dimensions other than the first have garbage / missing values (see the attached output file `test.nc`). In a case with 2 DEs per PET, the output file does *not* contain the replicated dimensions: these are removed in the function `IO::redist_arraycreate1de` (see the attached output file `test_multiple_des_per_pet.nc`). Trying to read the file into a new Array then fails with this error:

```
20230927 090845.524 ERROR            PET1 ESMCI_Array.C:6272 ESMCI::Array::tRedistStore() Invalid argument  - srcArray and dstArray must provide identical number of exclusive elements
20230927 090845.524 ERROR            PET1 ESMCI_Array.C:6052 ESMCI::Array::redistStore() Invalid argument  - Internal subroutine call returned Error
20230927 090845.524 ERROR            PET1 ESMCI_IO.C:396 ESMCI::IO::read() Invalid argument  - Internal subroutine call returned Error
20230927 090845.524 ERROR            PET1 ESMCI_IO.C:282 ESMCI::IO::read() Invalid argument  - Internal subroutine call returned Error
20230927 090845.524 ERROR            PET1 ESMCI_Array.C:3970 ESMCI::Array::read() Unable to read from file  - Internal subroutine call returned Error
20230927 090845.524 ERROR            PET1 ESMCI_Array_F.C:833 c_esmc_arrayread() Unable to read from file  - Internal subroutine call returned Error
20230927 090845.524 ERROR            PET1 ESMF_ArrayHa.F90:713 ESMF_ArrayRead Unable to read from file  - Internal subroutine call returned Error
20230927 090845.524 ERROR            PET1 replicatedDimsIO.F90:74 Unable to read from file  - Passing error in return code
```

So the problems I see are:
1. Output of Arrays with replicated dimensions is fundamentally different depending on whether the layout has 1 DE per PET or more than 1 DE per PET. (The output with 1 DE per PET, which maintains the replicated dimensions in the file, seems more intuitively correct to me, but I'm not sure.)
2. Both cases have issues: either garbage values or problems reading the file back in.

[test.nc.zip](https://github.com/esmf-org/esmf/files/12740375/test.nc.zip)
[test_multiple_des_per_pet.nc.zip](https://github.com/esmf-org/esmf/files/12740376/test_multiple_des_per_pet.nc.zip)

**Comment by billsacks** [comment]
Feeling from core team call: The right thing to do is probably to compress the replicated dimensions, as is currently done in the case with > 1 DE per PET. To accomplish this, I'll have the IO always do a redist if there are replicated dimensions. Then I'll try to either get the read working or abort with an informative error message if you try to do a read into an array with replicated dimensions.

**Comment by billsacks** [comment]
As mentioned in the previous comment, we decided that the right thing to do for I/O with replicated dimensions is probably to remove the replicated dimensions from the output file (rather than having a bunch of duplicated values or missing values in the output). The minimum diffs that I attempted to initially address the write piece of this issue – building on top of the changes in #187 – were:

```diff
diff --git a/src/Infrastructure/IO/src/ESMCI_IO.C b/src/Infrastructure/IO/src/ESMCI_IO.C
index d74e59329c..d582ae2f31 100644
--- a/src/Infrastructure/IO/src/ESMCI_IO.C
+++ b/src/Infrastructure/IO/src/ESMCI_IO.C
@@ -1219,7 +1219,16 @@ bool IO::redist_check(Array *array_p, int *rc) {
   if (ESMC_LogDefault.MsgFoundError(localrc, ESMCI_ERR_PASSTHRU, ESMC_CONTEXT, rc)) {
     return false;
   }
-  need_redist = de_test != 0;
+
+  // We do a redist in either of these cases:
+  // (1) Any PETs have a DE count other than 1
+  // (2) The array has replicated dimensions
+  //
+  // There may be a better way to deal with the replicated dimension case, rather than
+  // requiring a full redist in that situation (particularly if we allow for the
+  // possibility that we'll have 0 DEs on some PETs). But for now, for simplicity, we go
+  // through the same procedure in this case.
+  need_redist = (de_test != 0) || (array_p->getReplicatedDimCount() > 0);

   if (rc) *rc = ESMF_SUCCESS;
   return need_redist;
```

These diffs force going through `redist_arraycreate1de` whenever we have an array with replicated dimensions, which removes the replicated dimensions from the array.

I also considered removing the check that the index space is large enough to distribute across all PETs (i.e., that we have at least one DE per PET):
```diff
diff --git a/src/Infrastructure/IO/src/ESMCI_IO.C b/src/Infrastructure/IO/src/ESMCI_IO.C
index c653580ffe..d74e59329c 100644
--- a/src/Infrastructure/IO/src/ESMCI_IO.C
+++ b/src/Infrastructure/IO/src/ESMCI_IO.C
@@ -1283,12 +1283,6 @@ void IO::redist_arraycreate1de(Array *src_array_p, Array **dest_array_p, int pet
     distgridToArrayMap = &(distgridToArrayMapVec[0]);
   }

-  if ((maxIndexTile[0]-minIndexTile[0]+1)<petCount){
-    ESMC_LogDefault.MsgFoundError(ESMF_RC_INTNRL_BAD,
-      "Index space too small to be distributed across all PETs", ESMC_CONTEXT, rc);
-    return; // bail out
-  }
-
   ESMCI::InterArray<int> minIndexInterface((int*)minIndexTile, ndims-replicatedDims);
   ESMCI::InterArray<int> maxIndexInterface((int*)maxIndexTile, ndims-replicatedDims);
 #if 0
```

I thought that that might not be needed anymore – particularly with the changes I made last year for multi-tile I/O (which can handle the case of a given PET not having any DEs for a given tile of an Array). And it seemed especially likely that we'd run into this limitation for Arrays with replicated dimensions, because now we would have fewer dimensions to distribute across... and, in fact, we do run into this limitation in the existing unit tests of Array I/O with replicated dimensions.

I should note that, even in looking at tests of the baseline (i.e., before the above changes), I saw weird behavior; this can be seen in the current unit tests in `ESMF_ArrayIOUTest.F90`: different tests have different patterns of missing values in the output file. I'm not sure if the problem is with the initial filling of the Array with values or with the output of the Array. (However, I did see that filling all points with the same constant value resulted in the same pattern of missing values in the output, so the problem isn't related to the details of my `fillArray1d` subroutine – although I did notice that the `lbound` and `ubound` of `arrayToFillPtr` were different from what I would have expected in the cases with weird looking output.):

```
netcdf Array_repli1_first {
dimensions:
	myData_dim001 = 10 ;
	myData_dim002 = 3 ;
variables:
	double myData(myData_dim002, myData_dim001) ;
data:

 myData =
  1, _, 101, _, 201, _, 301, _, 401, 501,
  2, _, 102, _, 202, _, 302, _, 402, 502,
  3, _, 103, _, 203, _, 303, _, 403, 503 ;
}

netcdf Array_repli2_first {
dimensions:
	myData_dim001 = 10 ;
	myData_dim002 = 3 ;
variables:
	double myData(myData_dim002, myData_dim001) ;
data:

 myData =
  1, 2, 103, 104, 205, 206, 307, 308, 409, 510,
  _, _, _, _, _, _, _, _, _, _,
  _, _, _, _, _, _, _, _, _, _ ;
}

netcdf Array_repli1_second {
dimensions:
	myData_dim001 = 10 ;
	myData_dim002 = 3 ;
variables:
	double myData(myData_dim002, myData_dim001) ;
data:

 myData =
  1, _, _, _, _, _, _, _, _, _,
  102, _, _, _, _, _, _, _, _, _,
  203, _, _, _, _, _, _, _, _, _ ;
}

netcdf Array_repli2_second {
dimensions:
	myData_dim001 = 10 ;
	myData_dim002 = 3 ;
variables:
	double myData(myData_dim002, myData_dim001) ;
data:

 myData =
  1, 2, 3, 4, 5, 6, 7, 8, 9, 10,
  101, 102, 103, 104, 105, 106, 107, 108, 109, 110,
  201, 202, 203, 204, 205, 206, 207, 208, 209, 210 ;
}
```

After making the above code changes, and temporarily removing the Read tests (since, as noted above, Reads currently fail when removing replicated dimensions), I was getting non-deterministic failures: sometimes everything passes, but other times I get failures in some tests, with a different number of failures different times. Checking the output from one of the failures, it looks like there was a crash in a DistGridDestroy, but I'm not sure about that.

From a time when there were no failures, this was the netcdf output:

```
netcdf Array_repli1_first {
dimensions:
	myData_dim001 = 3 ;
variables:
	double myData(myData_dim001) ;
data:

 myData = 1, 0, 101 ;
}

netcdf Array_repli2_first {
dimensions:
	myData_dim001 = 10 ;
variables:
	double myData(myData_dim001) ;
data:

 myData = 207, 2, 103, 104, 205, 206, 307, 308, 409, 510 ;
}

netcdf Array_repli1_first_dim2 {
dimensions:
	myData_dim001 = 3 ;
variables:
	double myData(myData_dim001) ;
data:

 myData = 1, 0, 0 ;
}

netcdf Array_repli2_first_dim2 {
dimensions:
	myData_dim001 = 10 ;
variables:
	double myData(myData_dim001) ;
data:

 myData = 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 ;
}
```

So this looks weird, at least for some of the files. Furthermore, the output differed a bit when I ran the tests before removing the Reads; I'm not sure if this was an issue of broken Reads contaminating things or non-determinism in the redistribution, but I'm kind of guessing the latter.

I hypothesized that at least some of the issue may be due to having 0 DEs per PET on some PETs. To deal with this, I increased the size of the DistGrid used in the unit tests from `(10,3)` to `(24,12)`. (I also restored the error check forcing at least 1 DE per PET and confirmed that this error was no longer triggered with this increase in size of the DistGrid.) Doing this got rid of the intermittent failures (at least in 10 runs; note that the Reads were still removed). The outputs still looked weird, though:

```
netcdf Array_repli1_first {
dimensions:
	myData_dim001 = 12 ;
variables:
	double myData(myData_dim001) ;
data:

 myData = 102, 0, 0, 0, 101, 1, 0, 0, 302, 0, 0, 0 ;
}

netcdf Array_repli2_first {
dimensions:
	myData_dim001 = 24 ;
variables:
	double myData(myData_dim001) ;
data:

 myData = 2215, 2, 3, 214, 105, 213, 107, 108, 209, 210, 211, 212, 313, 314,
    315, 316, 417, 418, 419, 420, 521, 522, 523, 524 ;
}

netcdf Array_repli1_first_dim2 {
dimensions:
	myData_dim001 = 12 ;
variables:
	double myData(myData_dim001) ;
data:

 myData = 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 ;
}

netcdf Array_repli2_first_dim2 {
dimensions:
	myData_dim001 = 24 ;
variables:
	double myData(myData_dim001) ;
data:

 myData = 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
    20, 21, 22, 23, 24 ;
}
```

I reran and got different results – so there was some non-determinism in the output:

```
netcdf Array_repli1_first {
dimensions:
	myData_dim001 = 12 ;
variables:
	double myData(myData_dim001) ;
data:

 myData = 1, 0, 0, 0, 101, 0, 0, 0, 201, 0, 0, 0 ;
}

netcdf Array_repli2_first {
dimensions:
	myData_dim001 = 24 ;
variables:
	double myData(myData_dim001) ;
data:

 myData = 1, 2, 3, 4, 105, 106, 107, 108, 209, 210, 211, 212, 313, 314, 315,
    316, 417, 418, 419, 420, 521, 522, 523, 524 ;
}

netcdf Array_repli1_first_dim2 {
dimensions:
	myData_dim001 = 12 ;
variables:
	double myData(myData_dim001) ;
data:

 myData = 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 ;
}

netcdf Array_repli2_first_dim2 {
dimensions:
	myData_dim001 = 24 ;
variables:
	double myData(myData_dim001) ;
data:

 myData = 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
    20, 21, 22, 23, 24 ;
}
```

**So, in conclusion, the simple attempt to output Arrays with replicated dimensions by removing the replicated dimensions seems to have issues, at least in some configurations.** It's possible that some of the unit test setup is problematic, but in any case, it seemed like this would take significant time to investigate, and given that I/O with replicated dimensions is not being requested currently, we decided to drop this for now.

**Comment by billsacks** [comment]
I was thinking a bit about what's different about Reads vs. Writes that might be causing Reads to fail when we collapse the replicated dimensions. The two most obvious differences are:
1. Writes do a redist from the original array to a temporary array without replicated dimensions; Reads do the reverse
2. Writes potentially reuse an existing RouteHandle if a compatible one is found, whereas Reads don't do this RouteHandle caching

I wondered if (2) might also be contributing to some of the issues documented above with Writes... e.g., if `rhh->isCompatible` may be incorrect in this case. However, my understanding of the code (possibly wrong... maybe worth confirming by adding some prints in a test) is that the reuse of RouteHandles will only kick in if we're doing a single Write call with multiple objects. i.e., I don't think it will kick in based on earlier Write calls, so shouldn't kick in if we're doing a single ESMF_ArrayWrite, as is done in `ESMF_ArrayIOUTest.F90`.

So the likely explanation is that redists from an array without replicated dimensions to an array with replicated dimensions are not yet fully implemented.
